content/blog/using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application.md

```diff
@@ -0,0 +1,545 @@
+simpleMPI{0..7}:             line : information about the filename and linenumber for a given $pc
+simpleMPI{0..7}:
+```
+
+There are also other top level `cuda` commands other than `info`:
+
+```shell
+dbg all> cuda help
+simpleMPI{0..7}: Print or select the CUDA focus.
+simpleMPI{0..7}:
+simpleMPI{0..7}: List of cuda subcommands:
+simpleMPI{0..7}:
+simpleMPI{0..7}: cuda block -- Print or select the current CUDA block.
+simpleMPI{0..7}: cuda device -- Print or select the current CUDA device.
+simpleMPI{0..7}: cuda grid -- Print or select the current CUDA grid.
+simpleMPI{0..7}: cuda kernel -- Print or select the current CUDA kernel.
+simpleMPI{0..7}: cuda lane -- Print or select the current CUDA lane.
+simpleMPI{0..7}: cuda sm -- Print or select the current CUDA SM.
+simpleMPI{0..7}: cuda thread -- Print or select the current CUDA thread.
+simpleMPI{0..7}: cuda warp -- Print or select the current CUDA warp.
+simpleMPI{0..7}:
+simpleMPI{0..7}: Type "help cuda" followed by cuda subcommand name for full documentation.
+simpleMPI{0..7}: Type "apropos word" to search for commands related to "word".
+simpleMPI{0..7}: Type "apropos -v word" for full documentation of commands related to "word".
+simpleMPI{0..7}: Command name abbreviations are allowed if unambiguous.
+```
+
+See [the cuda-gdb documentation](https://docs.nvidia.com/cuda/cuda-gdb/index.html) for more information.
+
+#### Single Stepping
+
+The usual `step` and `next` commands are supported. In a GPU kernel, stepping
+has the granularity of a warp. A warp is a group of 32 GPU threads and is the
+granularity the GPU scheduler works at. See
+[the cuda-gdb documentation](https://docs.nvidia.com/cuda/cuda-gdb/index.html#single-stepping)
+for more info.
+
+```shell
+simpleMPI{0..7}: Breakpoint 1,  at simpleMPI.cu:55
+dbg all> list 55
+simpleMPI{0..7}: 50       }
+simpleMPI{0..7}: 51
+simpleMPI{0..7}: 52     // Device code
+simpleMPI{0..7}: 53     // Very simple GPU Kernel that computes square roots of input numbers
+simpleMPI{0..7}: 54     __global__ void simpleMPIKernel(float *input, float *output) {
+simpleMPI{0..7}: 55       int tid = blockIdx.x * blockDim.x + threadIdx.x;
+simpleMPI{0..7}: 56       output[tid] = sqrt(input[tid]);
+simpleMPI{0..7}: 57     }
+simpleMPI{0..7}: 58
+simpleMPI{0..7}: 59     // Initialize an array with random data (between 0 and 1)
+dbg all> next # set tid
+simpleMPI{0..7}: 56       output[tid] = sqrt(input[tid]);
+dbg all> p tid
+simpleMPI{0..7}: 64
+dbg all> p output[tid]
+simpleMPI{0..7}: 0
+dbg all> n
+simpleMPI{0..7}: 57     }
+dbg all> p output[tid]
+simpleMPI{0}: 0.516397
+simpleMPI{1}: 0.392356
+simpleMPI{2}: 0.883166
+simpleMPI{3}: 0.806327
+simpleMPI{4}: 0.458106
+simpleMPI{5}: 0.999046
+simpleMPI{6}: 0.366925
+simpleMPI{7}: 0.646477
+```
+
+## Next Steps
+
+You now know how to start debugging a multinode MPI/CUDA application with gdb4hpc in the HPE Cray Programming Environment!
```

Comment:
Is there any other benefit statement you can add? Like "This will save you hours of time in trying to find those little bugs."?